File: sdks/typescript/examples/team-sharing.ts
/**
 * Team-sharing example — Wave 3 APIs.
 *
 * Demonstrates the full team lifecycle:
 *   1. Create a team (caller becomes owner)
 *   2. Add a member by their user_id
 *   3. Store a memory into the team's scope via `shareWith`
 *   4. Search the team scope via `scope: "team:<uuid>"`
 *   5. Remove the member (optional cleanup)
 *
 * Run with:
 *   ENGRAM_API_KEY=pr_live_... npx tsx examples/team-sharing.ts
 */

import { EngramClient, EngramAPIError } from "../src/index.js";

async function main(): Promise<void> {
  const apiKey = process.env.ENGRAM_API_KEY;
  if (!apiKey) {
    throw new Error("Set ENGRAM_API_KEY first.");
  }

  const client = new EngramClient({ apiKey });

  // 1. Create the team. Slug must be 3-48 chars, lowercase alphanumerics
  //    and hyphens. The caller's own user becomes owner automatically.
  const team = await client.createTeam({
    name: "Platform Team",
    slug: `platform-${Date.now().toString(36)}`,
  });
  console.log("Team created:", team.id, team.slug);

  // 2. Add a second user. You need their Engram user UUID (not email).
  //    This call requires the caller to be owner or admin.
  const collaboratorId = process.env.COLLAB_USER_ID;
  if (collaboratorId) {
    try {
      const added = await client.addTeamMember(team.id, {
        userId: collaboratorId,
        role: "member",
      });
      console.log("Added:", added.user_id, "as", added.role);
    } catch (err) {
      if (err instanceof EngramAPIError && err.status === 404) {
        console.warn("User not found, skipping.");
      } else {
        throw err;
      }
    }
  }

  // 3. Store a memory that fans out to the team collection. The
  //    primary write still lands in the caller's personal collection;
  //    team fanout happens alongside with the same doc_id so it's
  //    addressable from either scope.
  const shared = await client.store({
    text: "Staging redis password rotates every 90 days — script in infra/rotate.sh.",
    category: "runbook",
    importance: 0.8,
    shareWith: [`team:${team.id}`],
  });
  console.log("Stored in personal + team:", shared.id);

  // 4. Search the team's collection specifically. Scope routes the
  //    search engine to the team's physical Qdrant collection and
  //    rechecks membership before returning anything.
  const teamHits = await client.search({
    query: "how do I rotate the redis password",
    scope: `team:${team.id}`,
    topK: 5,
  });
  console.log(`Team scope returned ${teamHits.results.length} result(s).`);

  // 5. Cleanup — remove the collaborator if we added one. Owners
  //    cannot remove themselves; for a full tear-down you'd delete
  //    the team entirely (not yet exposed in this SDK surface).
  if (collaboratorId) {
    const removed = await client.removeTeamMember(team.id, collaboratorId);
    console.log("Removed:", removed.user_id, "=>", removed.removed);
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: sdks/python/examples/team_sharing.py
"""End-to-end team sharing flow.

Creates a team, adds a member, stores a memory into it, and then
searches within the team scope. The target user id is read from
argv[1] so the script doesn't hard-code another account.

    export ENGRAM_API_KEY=pr_live_...
    python examples/team_sharing.py <invitee_user_uuid>

If the invitee hasn't signed up yet, the add step returns a 404 —
the SDK surfaces it as ``EngramAPIError`` with status 404 and the
script prints a readable message before exiting.
"""

from __future__ import annotations

import sys

from engram import (
    EngramAPIError,
    EngramClient,
    EngramError,
)


def main(argv: list[str]) -> int:
    if len(argv) < 2:
        print("usage: python team_sharing.py <invitee_user_uuid>", file=sys.stderr)
        return 2
    invitee = argv[1]

    client = EngramClient()
    try:
        # 1. Create a team. Slug must be lowercase alphanumeric + hyphens.
        team = client.create_team(name="Platform Ops", slug="platform-ops-demo")
        print("[create]  team={} slug={}".format(team.id, team.slug))

        # 2. Invite a second human.
        try:
            client.add_team_member(team.id, user_id=invitee, role="member")
            print("[add]     member={}".format(invitee))
        except EngramAPIError as exc:
            if exc.status_code == 404:
                print(
                    "Invitee {} not found — they need to sign up first.".format(invitee),
                    file=sys.stderr,
                )
                return 1
            raise

        # 3. Store into both the personal collection and the team one
        #    in a single call. The cloud fans the write out atomically —
        #    if the team write fails the personal write rolls back too.
        stored = client.store(
            "Our primary queue is SQS FIFO as of 2025-Q3",
            category="decisions",
            importance=0.8,
            share_with=["team:{}".format(team.id)],
        )
        print("[store]   id={} status={}".format(stored.id, stored.status))

        # 4. Search within the team scope — should return the memory we
        #    just shared.
        scope = "team:{}".format(team.id)
        hits = client.search(
            "what messaging system does the platform team use",
            top_k=5,
            scope=scope,
        )
        print("[search]  scope={} -> {} hits".format(scope, len(hits.results)))
        for i, hit in enumerate(hits.results, start=1):
            print("  {}. {:.2f}  {}".format(i, hit.score, hit.text[:100]))

        # 5. List every team the caller is in, just to sanity-check.
        teams = client.list_teams()
        print("[list]    caller is a member of {} teams".format(len(teams)))
        for t in teams:
            print("  - {} ({}) role={}".format(t.name, t.slug, t.role or "?"))
    except EngramError as exc:
        print("Engram call failed: {}".format(exc), file=sys.stderr)
        return 1
    finally:
        client.close()

    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv))
